hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/SnapshotChainManager.java

```diff
@@ -0,0 +1,446 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.hadoop.ozone.om;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.hadoop.hdds.utils.db.Table;
+import org.apache.hadoop.hdds.utils.db.TableIterator;
+import org.apache.hadoop.ozone.om.helpers.SnapshotInfo;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.LinkedHashMap;
+import java.util.TreeMap;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is used for creating and accessing Snapshot Chains.
+ *
+ * The snapshot chain maintains the in-memory sequence of snapshots
+ * created in chronological order.  There are two such snapshots maintained
+ * i.) Path based snapshot chain, sequence of snapshots created for a
+ * given /volume/bucket
+ * ii.) Global snapshot chain, sequence of all snapshots created in order
+ *
+ * On start, the snapshot chains are initialized from the on disk
+ * SnapshotInfoTable from the om RocksDB.
+ */
+public class SnapshotChainManager {
+  private LinkedHashMap<String, SnapshotChainInfo>  snapshotChainGlobal;
+  private Map<String, LinkedHashMap<String, SnapshotChainInfo>>
+      snapshotChainPath;
+  private Map<String, String> latestPathSnapshotID;
+  private String latestGlobalSnapshotID;
+  private static final Logger LOG =
+      LoggerFactory.getLogger(SnapshotChainManager.class);
+
+  public SnapshotChainManager(OMMetadataManager metadataManager)
+          throws IOException {
+    snapshotChainGlobal = new LinkedHashMap<>();
+    snapshotChainPath = new HashMap<>();
+    latestPathSnapshotID = new HashMap<>();
+    latestGlobalSnapshotID = null;
+    loadFromSnapshotInfoTable(metadataManager);
+  }
+
+  private void addSnapshotGlobal(String snapshotID,
+                            String prevGlobalID) throws IOException {
+    // set previous snapshotID to null if it is "" for
+    // internal in-mem structure
+    if (prevGlobalID != null && prevGlobalID.equals("")) {
+      prevGlobalID = null;
+    }
+    // on add snapshot; set previous snapshot entry nextSnapshotID =
+    // snapshotID
+    if (prevGlobalID != null &&
+        snapshotChainGlobal.containsKey(prevGlobalID)) {
+      snapshotChainGlobal
+          .get(prevGlobalID)
+          .setNextSnapshotID(snapshotID);
+    }
+    if (prevGlobalID != null &&
+        !snapshotChainGlobal.containsKey(prevGlobalID)) {
+      throw new IOException("Snapshot Chain corruption: "
+          + " previous snapshotID given but no associated snapshot "
+      + "found in snapshot chain: SnapshotID "
+      + prevGlobalID);
+    }
+    snapshotChainGlobal.put(snapshotID,
+            new SnapshotChainInfo(snapshotID, prevGlobalID, null));
+
+    // set state variable latestGlobal snapshot entry to this snapshotID
+    latestGlobalSnapshotID = snapshotID;
```

Comment:
There is a single `SnapshotChainManager` instance with the `OmMetadataManager` / `SnapshotMetadataManager.`  Within the metadata manager, it will add/modify snapshots to both the `SnapshotInfo `table and snapshot chain.  For all snapshot operations it will ensure the `SnapshotInfo` table is consistent with the snapshot chain - we can create synchronized blocks or similar for that.  This can be applied by the metadata manager for consistency between the table and chain and across threads.